### PR TITLE
JAMES-2813 Add first tasks serialization

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

**Comment by trantienduchn** [comment]
```
Progress (11): 0.5/12 MB | 0.2/2.4 MB | 0.2/3.2[ERROR] src/main/java/org/apache/james/json/DTOModule.java:[32,54] (naming) InterfaceTypeParameterName: Interface type name 'ModuleType' must match pattern '(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)'.
[58a14e33068b480ab0254609d2fae447209157cd]  MB | 212/48[ERROR] src/main/java/org/apache/james/json/DTOModule.java:[94,33] (naming) MethodTypeParameterName: Method type name 'ModuleType' must match pattern '(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)'.
```

**Comment by Arsnael** [comment]
```
[ERROR] src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/dto/EventDTOModule.java:[27,20] (naming) MethodTypeParameterName: Method type name 'EventType' must match pattern '(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)'.
```

File: json/src/test/java/org/apache/JsonGenericSerializerTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                TestModules.FIRST_TYPE,
+                TestModules.SECOND_TYPE);
+
+        assertThatJson(
+                serializer.serialize(SECOND))
+            .isEqualTo(SECOND_JSON);
+
+        assertThatJson(
+                serializer.serialize(FIRST))
+            .isEqualTo(FIRST_JSON);
```
Comment (Arsnael) [inline json/src/test/java/org/apache/JsonGenericSerializerTest.java]: I thought the rule was one assert for one test?

File: json/src/test/java/org/apache/JsonGenericSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.james.json.DTO;
+import org.apache.james.json.JsonGenericSerializer;
+import org.junit.jupiter.api.Test;
+
+class JsonGenericSerializerTest {
+    FirstDomainObject FIRST = new FirstDomainObject(Optional.of(1L), ZonedDateTime.parse("2016-04-03T02:01+07:00[Asia/Vientiane]"), "first payload");
+    SecondDomainObject SECOND = new SecondDomainObject(UUID.fromString("4a2c853f-7ffc-4ce3-9410-a47e85b3b741"), "second payload");
+
+    String FIRST_JSON = "{\"type\":\"first\",\"id\":1,\"time\":\"2016-04-03T02:01+07:00[Asia/Vientiane]\",\"payload\":\"first payload\"}";
+    String SECOND_JSON = "{\"type\":\"second\",\"id\":\"4a2c853f-7ffc-4ce3-9410-a47e85b3b741\",\"payload\":\"second payload\"}";
```
Comment (Arsnael) [inline json/src/test/java/org/apache/JsonGenericSerializerTest.java]: Why the variables are not `private static final`?

File: json/src/test/java/org/apache/dto/TestModules.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.dto;
+
+import org.apache.james.json.DTOModule;
+
+public interface TestModules {
+
+
```
Comment (Arsnael) [inline json/src/test/java/org/apache/dto/TestModules.java]: remove extra line break

File: event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java [hunk 4]
```diff
@@ -47,7 +47,7 @@ public UnknownEventException(JsonGenericSerializer.UnknownTypeException original
         }
     }
```
Comment (Arsnael) [inline event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java]: Commit message -> s/DTOMdule/DTOModule

File: json/src/main/java/org/apache/james/json/DTOModule.java [hunk 5]
```diff
@@ -41,61 +45,81 @@ public Builder(Class<T> type) {
...
+
+                        public <ModuleTypeT extends DTOModule<T, U>> ModuleTypeT withFactory(ModuleFactory<T, U, ModuleTypeT> factory) {
+                            return factory.create(converter, toDomainObjectConverter, type, dtoType, typeName);
+                        }
                     }
                 }
             }
         }
+
+
```
Comment (Arsnael) [inline json/src/main/java/org/apache/james/json/DTOModule.java]: Why all those extra line breaks?

File: server/task-json/src/test/java/org/apache/james/server/task/json/TaskSerializationTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+public class TaskSerializationTest {
+
+    public static final String SERIALIZED_FAILED_TASK = "{\"type\": \"failed-task\"}";
+    private TaskDTOModule failedTaskModule = DTOModule
+            .forDomainObject(FailedTask.class)
+            .convertToDTO(FailedTaskDTO.class)
+            .toDomainObjectConverter(dto -> new FailedTask())
+            .toDTOConverter((task, typeName) -> new FailedTaskDTO(typeName))
+            .typeName("failed-task")
+            .withFactory(TaskDTOModule::new);
```
Comment (Arsnael) [inline server/task-json/src/test/java/org/apache/james/server/task/json/TaskSerializationTest.java]: The indent is different between the different `TaskDTOModule` declarations

File: server/task-json/src/test/java/org/apache/james/server/task/json/TaskSerializationTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            .forDomainObject(ThrowingTask.class)
+            .convertToDTO(ThrowingTaskDTO.class)
+            .toDomainObjectConverter(dto -> new ThrowingTask())
+            .toDTOConverter((task, typeName) -> new ThrowingTaskDTO(typeName))
+            .typeName("throwing-task")
+            .withFactory(TaskDTOModule::new);
+    
+    @BeforeEach
+    void setUp() {
+    }
```
Comment (Arsnael) [inline server/task-json/src/test/java/org/apache/james/server/task/json/TaskSerializationTest.java]: If it's empty it's not useless to define it?

File: server/task/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java [hunk 8]
```diff
@@ -38,13 +38,13 @@
 
 class SerialTaskManagerWorkerTest {
 
+
```
Comment (Arsnael) [inline server/task/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java]: Why extra line breaks everywhere?

File: server/task/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java [hunk 9]
```diff
@@ -38,13 +38,13 @@
...
     private final SerialTaskManagerWorker worker = new SerialTaskManagerWorker();
 
-    private final Task successfulTask = () -> Task.Result.COMPLETED;
-    private final Task failedTask = () -> Task.Result.PARTIAL;
-    private final Task throwingTask = () -> {
-        throw new RuntimeException("Throwing Task");
-    };
+
```
Comment (Arsnael) [inline server/task/src/test/java/org/apache/james/task/SerialTaskManagerWorkerTest.java]: Idem

File: server/task-json/src/main/java/org/apache/james/server/task/json/TaskDeserializer.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.james.task.Task;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class TaskDeserializer {
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/task-json/src/main/java/org/apache/james/server/task/json/TaskDeserializer.java]: So seems those in `mailbox/tools/indexer` got removed in the last commit. Was it necessary to have those temporarily for the dev, or was just it just a mistake that got forgotten a bit in the middle? I'm trying to understand here...

**Comment by Arsnael** [comment]
FYI: The first batch of outdated comments are because there was a force pushed rebase that happened after I reviewed the first commit, but I don't believe it changed during that event

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/TaskSerializerTest.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.james.mailbox.model.TestId;
+import org.junit.jupiter.api.Test;
+
+
+class TaskSerializerTest {
+
+    public static final String TASK_AS_STRING = "{" +
+            "\"type\": \"mailboxReIndexing\"," +
+            "\"parameters\": {\"mailboxId\": \"1\"}" +
+            "}";
```
[the remark this answers is not in the document]
Comment (blackheaven) [reply mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/TaskSerializerTest.java]: probably outdated

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/UnsupportedTypeException.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.mailbox.tools.indexer;
+
+public class UnsupportedTypeException extends RuntimeException {
+    private final String type;
+
+  public UnsupportedTypeException(String type) {
```
[the remark this answers is not in the document]
Comment (blackheaven) [reply mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/UnsupportedTypeException.java]: removed

File: server/task-json/src/main/java/org/apache/james/server/task/json/TaskDeserializer.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        return getFactory(taskAsJson).create(parameters);
+    }
+
+    private Factory getFactory(JsonNode taskAsJson) {
+        String type = Optional.ofNullable(taskAsJson.get("type"))
+            .map(JsonNode::asText)
+            .orElseThrow(() -> new InvalidTaskException());
+        return Optional.ofNullable(registry.get(type))
+            .orElseThrow(() -> new UnsupportedTypeException(type));
+
```
[the remark this answers is not in the document]
Comment (blackheaven) [reply server/task-json/src/main/java/org/apache/james/server/task/json/TaskDeserializer.java]: removed

**Pull request closed without merging**